compiler/passes/normalize.cpp

```diff
@@ -978,6 +944,43 @@ static TypeSymbol* resolveTypeAlias(SymExpr* se)
   return NULL;
 }
 
+//
+// These RiSpec helpers implement reduce intents of the form
+// type(someArg), see e.g. test/parallel/forall/vass/3types-*.
+// We want to keep these reduce intents in their original form
+// until reduce-intent handling later, presently in resolution.
+//
+// We do it here to avoid transforming it into _type_construct_C ( ... ).
+// That would be incorrect because this is a special syntax for reduce intent.
+//
+static SymExpr* callUsedInRiSpec(Expr* call, CallExpr* parent) {
+  if (parent && parent->isPrimitive(PRIM_MOVE)) {
+    SymExpr* destSE = toSymExpr(parent->get(1));
+    Symbol* dest = destSE->symbol();
+    SymExpr* riSpecMaybe = dest->firstSymExpr();
+    if (ForallStmt* fs = toForallStmt(riSpecMaybe->parentExpr)) {
+      for_vector(Expr, riSpec, fs->withClause()->riSpecs)
+        if (riSpecMaybe == riSpec)
+          return riSpecMaybe;
+    }
+  }
+  return NULL;
+}
+static void restoreRiSpecCall(SymExpr* riSpec, CallExpr* call) {
```

Comment:
I think it'd be nice to have a comment here pointing up and saying that this function partially un-does the work of call_constructor_for_class so that reduce intents don't get messed up.